README.rst
Policyd rate limit
==================


.. image:: https://img.shields.io/pypi/v/policyd-rate-limit.svg
    :target: https://pypi.python.org/pypi/policyd-rate-limit

.. image:: https://img.shields.io/pypi/l/policyd-rate-limit.svg
    :target: https://www.gnu.org/licenses/gpl-3.0.html

Postfix policyd server allowing to limit the number of mails accepted by
postfix over several time periods, by sasl usernames and/or ip addresses.


Installation
------------

First, create the user that will run the daemon and its home directory::

    mkdir -p /var/lib/policyd-rate-limit
    useradd policyd-rate-limit -d /var/lib/policyd-rate-limit

Depending of the backend storage you planning to use, you may need to install additional packages.
(The default settings use the sqlite3 bakends and do not need extra packages).

* `mysqldb <https://github.com/PyMySQL/mysqlclient-python>`_
  (``sudo apt-get install python3-mysqldb`` on debian like systems) for the mysql backend.
* `psycopg2 <https://pypi.python.org/pypi/psycopg2>`_
  (``sudo apt-get install python3-psycopg2`` on debian like systems) fot the postgreysql backend

Install with pip::

    sudo pip3 install policyd-rate-limit

or from source code::

    sudo make install

This will install the ``policyd_rate_limit`` module, the ``policyd-rate-limit`` binary,
copy the default config to ``/etc/policyd-rate-limit.conf`` if the file do not exists,
copy an init script to ``/etc/init.d/policyd-rate-limit`` and an unit file to
``/etc/systemd/system/policyd-rate-limit.service``.

After the installation, you may need to run ``sudo systemctl daemon-reload`` for make the unit
file visible by systemd.

You should run ``policyd-rate-limit --clean`` on a regular basis to delete old records from the
database. It could be wise to put it in a daily cron, for example::

    0 0 * * * policyd-rate-limit /usr/local/bin/policyd-rate-limit --clean >/dev/null

Settings
--------

``policyd-rate-limit`` search for its config first in ``~/.config/policyd-rate-limit.conf``
If not found, then in ``/etc/policyd-rate-limit.conf``, and if not found use the default config.

* ``debug``: make ``policyd-rate-limit`` output logs to stderr.
  The default is True.
* ``user``: The user ``policyd-rate-limit`` will use to drop privileges.
  The default is ``"policyd-rate-limit"``.
* ``group``: The group ``policyd-rate-limit`` will use to drop privileges.
  The defaut is ``"policyd-rate-limit"``.
* ``pidfile``: path where the program will try to write its pid to.
  The default is ``"/var/run/policyd-rate-limit/policyd-rate-limit.pid"``.
  ``policyd-rate-limit`` will try to create the parent directory and chown it if it do not exists.
* ``mysql_config``: The config to connect to a mysql server
* ``pgsql_config``: The config to connect to a postgreysql server
* ``sqlite_config``: The config to connect to a sqlite3 database.
* ``backend``: Which data backend to use. Possible values are ``0`` for sqlite3, ``1`` for mysql
  and ``2`` for postgreysql. The default is ``0``, use the sqlite3 backend.
* ``SOCKET``: The socket to bind to. Can be a path to an unix socket or a couple (ip, port).
  The default is ``"/var/spool/postfix/ratelimit/policy"``.
  ``policyd-rate-limit`` will try to create the parent directory and chown it if it do not exists.
* ``socket_permission``: Permissions on the unix socket (if unix socket used).
  The default is ``0o666``.
* ``limits``: A list of couple (number of emails, number of seconds). If one of the element of the
  list is exeeded (more than 'number of emails' on 'number of seconds' for an ip address or an sasl
  username), postfix will return a temporary failure.
* ``limit_by_sasl``: Apply limits by sasl usernames. The default is ``True``.
* ``limit_by_ip``: Apply limits by ip addresses. The default is ``False``.
* ``limited_netword``: A list of ip networks in cidr notation on which limits are applied. An empty
  list is equal to ``limit_by_ip = False``, put ``"0.0.0.0/0"`` and ``::/0`` for every ip addresses.
* ``success_action``: If not limits are reach, which action postfix should do. The default is
  ``"dunno"``. See http://www.postfix.org/access.5.html for possible actions.
* ``fail_action``: If a limit is reach, which action postfix should do. The default is
  ``"defer_if_permit Rate limit reach, retry later"``.
  See http://www.postfix.org/access.5.html for possible actions.
* ``config_file``: This parameter is automatically set to the path of the configuration file
  currently in use. You can call it conjunction with **--get-config** to known which configuration
  file is used.


* ``report``: if ``True``, send a report to ``report_to`` about users reaching limits each time
  --clean is called. The default is ``False``.
* ``report_from``: From who to send emails reports. It must be defined when ``report`` is ``True``.
* ``report_to``: Address to send emails reports to. It must be defined when ``report`` is ``True``.
* ``report_subject``: Subject of the report email. The default is ``"policyd-rate-limit report"``.
* ``report_limits``: List of number of seconds from the limits list for which you want to be reported.
  The default is ``[86400]``.
* ``report_only_if_needed``: Only send a report if some users have reach a reported limit.
  The default is ``True``.


* ``smtp_server``: The smtp server to use to send emails ``("host", port)``.
  The default is ``("localhost", 25)``.
* ``smtp_starttls``: Should we use starttls to send mails ? (you should set this to ``True`` if
  you use ``smtp_credentials``). The default is ``False``.
* ``smtp_credentials``: Should we use credentials to connect to smtp_server ?
  if yes set ``("user", "password")``, else ``None``. The default is ``None``.


Postfix settings
----------------

For postfix 3.0 and later I recommend using the example below. It ensure that if policyd-rate-limit
become unavailable for any reason, postfix will ignore it and keep accepting mail as if the rule
was not here. I find it nice has in my opinion, policyd-rate-limit is a "non-critical" policy
service.

    /etc/postfix/main.cf::

        smtpd_recipient_restrictions =
            ...,
            check_policy_service { unix:ratelimit/policy, default_action=DUNNO },
            ...


On previous postfix versions, you must use:

    /etc/postfix/main.cf::

        smtpd_recipient_restrictions =
            ...,
            check_policy_service unix:ratelimit/policy,
            ...


debian/changelog
policyd-rate-limit (0.5.1-1) unstable; urgency=low

  * Initial release. (Closes: #831456)
  * Disable the usage of upstream init scripts (patch ignore-init-service.patch)
    and use policyd-rate-limit.service and policyd-rate-limit.init instead.
    They are identical to the upstream ones except for /usr/local/bin/ replaced
    by /usr/bin/. It also let dh_installdeb automatically setup symbolic links
    in rc* in postinst and postrm.

 -- Valentin Samir <valentin.samir@crans.org>  Mon, 18 Jul 2016 20:19:00 +0200


policyd_rate_limit/utils.py
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License version 3 for
# more details.
#
# You should have received a copy of the GNU General Public License version 3
# along with this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# (c) 2015-2016 Valentin Samir
# -*- mode: python; coding: utf-8 -*-
import os
import threading
import collections
import ipaddress
import time
import imp
import pwd
import grp
import warnings
import smtplib
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart

from policyd_rate_limit.const import SQLITE_DB, MYSQL_DB, PGSQL_DB
from policyd_rate_limit import config as default_config


def ip_network(ip_str):
    try:
        return ipaddress.IPv4Network(ip_str)
    except ipaddress.AddressValueError:
        return ipaddress.IPv6Network(ip_str)


class Config(object):
    """Act as a config module, missing parameters fallbacks to default_config"""
    def __init__(self, config_file=None):
        if config_file is None:
            # search for config files in the following locations
            config_files = [
                os.path.expanduser("~/.config/policyd-rate-limit.conf"),
                "/etc/policyd-rate-limit.conf",
            ]
        else:
            config_files = [config_file]
        for config_file in config_files:
            if os.path.isfile(config_file):
                try:
                    self._config = imp.load_source('config', config_file)
                    self.config_file = config_file
                    break
                except PermissionError:
                    pass
        # if not config file found, raise en error
        else:
            raise ValueError(
                "No config file found or bad permissions, searched for %s" % (
                    ", ".join(config_files),
                )
            )

        self.limited_netword = [ip_network(net) for net in self.limited_netword]

    def __getattr__(self, name):
        try:
            return getattr(self._config, name)
        # If an parameter is not defined in the config file, return its default value.
        except AttributeError:
            return getattr(default_config, name)


class LazyConfig(object):
    """A lazy proxy to the Config class allowing to import config before it is initialized"""
    _config = None
    format_str = None

    def __getattr__(self, name):
        if self._config is None:
            raise RuntimeError("config is not initialized")
        return getattr(self._config, name)

    def setup(self, config_file=None):
        """initialize the config"""
        global cursor
        # initialize config
        self._config = Config(config_file)

        # make the cursor class function of the configured backend
        if config.backend == SQLITE_DB:
            cursor = make_cursor("sqlite_cursor", config.backend, config.sqlite_config)
            self.format_str = "?"
        elif config.backend == MYSQL_DB:
            cursor = make_cursor("mysql_cursor", config.backend, config.mysql_config)
            self.format_str = "%s"
        elif config.backend == PGSQL_DB:
            cursor = make_cursor("pgsql_cursor", config.backend, config.pgsql_config)
            self.format_str = "%s"
        else:
            raise ValueError("backend %s unknown" % config.backend)


def make_directories():
    """Create directory for pid and socket and chown if needed"""
    try:
        uid = pwd.getpwnam(config.user).pw_uid
    except KeyError:
        raise ValueError("User %s in config do not exists" % config.user)
    try:
        gid = grp.getgrnam(config.group).gr_gid
    except KeyError:
        raise ValueError("Group %s in config do not exists" % config.group)
    pid_dir = os.path.dirname(config.pidfile)
    if not os.path.exists(pid_dir):
        os.mkdir(pid_dir)
    if not os.listdir(pid_dir):
        os.chmod(pid_dir, 0o755)
        os.chown(pid_dir, uid, gid)
    if isinstance(config.SOCKET, str):
        socket_dir = os.path.dirname(config.SOCKET)
        if not os.path.exists(socket_dir):
            os.mkdir(socket_dir)
        if not os.listdir(socket_dir):
            os.chown(socket_dir, uid, gid)
    if config.backend == SQLITE_DB:
        try:
            db_dir = os.path.dirname(config.sqlite_config["database"])
            if not os.path.exists(db_dir):
                os.mkdir(db_dir)
            if not os.listdir(db_dir):
                os.chmod(db_dir, 0o700)
                os.chown(db_dir, uid, gid)
        except KeyError:
            pass


def drop_privileges():
    """If current running use is root, drop privileges to user and group set in the config"""
    if os.getuid() != 0:
        # We're not root so, like, whatever dude
        return

    # Get the uid/gid from the name
    running_uid = pwd.getpwnam(config.user).pw_uid
    running_gid = grp.getgrnam(config.group).gr_gid

    # Remove group privileges
    os.setgroups([])

    # Try setting the new uid/gid
    os.setgid(running_gid)
    os.setuid(running_uid)

    # Ensure a very conservative umask
    os.umask(0o077)


def make_cursor(name, backend, config):
    """Create a cursor class usable as a context manager, binding to the backend selected"""
    if backend == MYSQL_DB:
        try:
            import MySQLdb
        except ImportError:
            raise ValueError(
                "You need to install the python3 module MySQLdb to use the mysql backend"
            )
        methods = {
            '_db': collections.defaultdict(lambda: MySQLdb.connect(**config)),
            'backend': MYSQL_DB,
            'backend_module': MySQLdb,
        }
    elif backend == SQLITE_DB:
        import sqlite3
        methods = {
            '_db': collections.defaultdict(lambda: sqlite3.connect(**config)),
            'backend': SQLITE_DB,
            'backend_module': sqlite3,
        }
    elif backend == PGSQL_DB:
        try:
            import psycopg2
        except ImportError:
            raise ValueError(
                "You need to install the python3 module psycopg2 to use the postgresql backend"
            )
        methods = {
            '_db': collections.defaultdict(lambda: psycopg2.connect(**config)),
            'backend': PGSQL_DB,
            'backend_module': psycopg2,
        }
    else:
        raise RuntimeError("backend %s unknown" % backend)
    newclass = type(name, (_cursor,), methods)
    return newclass


class _cursor(object):
    """cursor template class"""
    backend = None
    backend_module = None

    @classmethod
    def get_db(cls):
        return cls._db[threading.current_thread()]

    @classmethod
    def set_db(cls, value):
        cls._db[threading.current_thread()] = value

    @classmethod
    def del_db(cls):
        try:
            cls._db[threading.current_thread()].close()
        except:
            pass
        try:
            del cls._db[threading.current_thread()]
        except KeyError:
            pass

    def __enter__(self):
        self.cur = self.get_db().cursor()
        if self.backend in [MYSQL_DB, PGSQL_DB]:
            try:
                if self.backend == MYSQL_DB:
                    self.cur.execute("DO 0")
                else:
                    self.cur.execute("SELECT 0")
                    self.cur.fetchone()
            except self.backend_module.Error as error:
                # SQL server has gone away, probably a timeout
                if error.args[0] in [2006, 8000, 8003, 8006]:
                    self.del_db()
                    self.cur.close()
                    self.cur = self.get_db().cursor()
        return self.cur

    def __exit__(self, exc_type, exc_value, traceback):
        self.cur.close()
        self.get_db().commit()


def is_ip_limited(ip):
    """Check if ``ip`` is part of a network of ``config.limited_netword``"""
    try:
        ip = ipaddress.IPv4Address(ip)
    except ipaddress.AddressValueError:
        ip = ipaddress.IPv6Address(ip)
    for net in config.limited_netword:
        if ip in net:
            return True
    return False


def print_fw(msg, length, filler=' ', align_left=True):
    msg = "%s" % msg
    if len(msg) > length:
        raise ValueError("%r must not be longer than %s" % (msg, length))
    if align_left:
        return "%s%s" % (msg, filler * (length - len(msg)))
    else:
        return "%s%s" % (filler * (length - len(msg)), msg)


def clean():
    """Delete old records from the database"""
    max_delta = 0
    for nb, delta in config.limits:
        max_delta = max(max_delta, delta)
    # remove old record older than 2*max_delta
    expired = int(time.time() - max_delta - max_delta)
    with cursor() as cur:
        cur.execute("DELETE FROM mail_count WHERE date <= %s" % config.format_str, (expired,))
        print("%d records deleted" % cur.rowcount)
        # if report is True, generate a mail report
        if config.report and config.report_to:
            send_report(cur)
            # The mail report has been successfully send, flush limit_report
            cur.execute("DELETE FROM limit_report")

    try:
        if config.backend == PGSQL_DB:
            # setting autocommit to True disable the transations. This is needed to run VACUUM
            cursor.get_db().autocommit = True
        with cursor() as cur:
            if config.backend == PGSQL_DB:
                cur.execute("VACUUM ANALYZE")
            elif config.backend == SQLITE_DB:
                cur.execute("VACUUM")
            elif config.backend == MYSQL_DB:
                if config.report:
                    cur.execute("OPTIMIZE TABLE mail_count, limit_report")
                else:
                    cur.execute("OPTIMIZE TABLE mail_count")
    finally:
        if config.backend == PGSQL_DB:
            cursor.get_db().autocommit = False


def send_report(cur):
    cur.execute("SELECT id, delta, hit FROM limit_report")
    # list to sort ids by hits
    report = list(cur.fetchall())
    if not config.report_only_if_needed or report:
        if report:
            text = ["Below is the table of users who hit a limit since the last cleanup:", ""]
            # dist to groups deltas by ids
            report_d = collections.defaultdict(list)
            max_d = {'id': 2, 'delta': 5, 'hit': 3}
            for (id, delta, hit) in report:
                report_d[id].append((delta, hit))
                max_d['id'] = max(max_d['id'], len(id))
                max_d['delta'] = max(max_d['delta'], len(str(delta)))
                max_d['hit'] = max(max_d['hit'], len(str(hit)))
            # sort by hits
            report.sort(key=lambda x: x[2])
            # table header
            text.append(
                "|%s|%s|%s|" % (
                    print_fw("id", max_d['id']),
                    print_fw("delta", max_d['delta']),
                    print_fw("hit", max_d['hit'])
                )
            )
            # table header/data separation
            text.append(
                "|%s+%s+%s|" % (
                    print_fw("", max_d['id'], filler='-'),
                    print_fw("", max_d['delta'], filler='-'),
                    print_fw("", max_d['hit'], filler='-')
                )
            )

            for (id, _, _) in report:
                # sort by delta
                report_d[id].sort()
                for (delta, hit) in report_d[id]:
                    # add a table row
                    text.append(
                        "|%s|%s|%s|" % (
                            print_fw(id, max_d['id'], align_left=False),
                            print_fw("%ss" % delta, max_d['delta'], align_left=False),
                            print_fw(hit, max_d['hit'], align_left=False)
                        )
                    )
        else:
            text = ["No user hit a limit since the last cleanup"]
        text.extend(["", "-- ", "policyd-rate-limit"])

        # Start building the mail report
        msg = MIMEMultipart()
        msg['Subject'] = config.report_subject or ""
        msg['From'] = config.report_from or ""
        msg['To'] = config.report_to
        msg.attach(MIMEText("\n".join(text), 'plain'))

        # check that smtp_server is wekk formated
        if isinstance(config.smtp_server, tuple):
            if len(config.smtp_server) >= 2:
                server = smtplib.SMTP(config.smtp_server[0], config.smtp_server[1])
            elif len(config.smtp_server) == 1:
                server = smtplib.SMTP(config.smtp_server[0], 25)
            else:
                raise ValueError("bad smtp_server should be a tuple (server_adress, port)")
        else:
            raise ValueError("bad smtp_server should be a tuple (server_adress, port)")

        try:
            # should we use starttls ?
            if config.smtp_starttls:
                server.starttls()
            # should we use credentials ?
            if config.smtp_credentials:
                if isinstance(config.smtp_credentials, tuple) and len(config.smtp_credentials) >= 2:
                    server.login(config.smtp_credentials[0], config.smtp_credentials[1])
                else:
                    ValueError("bad smtp_credentials should be a tuple (login, password)")
            server.sendmail(config.report_from or "", config.report_to, msg.as_string())
        finally:
            server.quit()


def database_init():
    """Initialize database (create the table and index)"""
    with cursor() as cur:
        query = """CREATE TABLE IF NOT EXISTS mail_count (
      id varchar(40) NOT NULL,
      date bigint NOT NULL
    );"""
        # if report is enable, also create the table for storing report datas
        if config.report:
            query_report = """CREATE TABLE IF NOT EXISTS limit_report (
      id varchar(40) NOT NULL,
      delta int NOT NULL,
      hit int NOT NULL DEFAULT 0
    );"""
        try:
            if cursor.backend == MYSQL_DB:
                # ignore possible warnings about the table already existing
                warnings.filterwarnings('ignore', category=cursor.backend_module.Warning)
            cur.execute(query)
            if config.report:
                cur.execute(query_report)
        finally:
            warnings.resetwarnings()
        try:
            cur.execute('CREATE INDEX mail_count_index ON mail_count(id, date)')
        except cursor.backend_module.Error as error:
            if error.args[0] == 'index mail_count_index already exists':
                pass
        # if report is enable, create and unique index on (id, delta)
        if config.report:
            try:
                cur.execute('CREATE UNIQUE INDEX limit_report_index ON limit_report(id, delta)')
            except cursor.backend_module.Error as error:
                if error.args[0] == 'index limit_report_index already exists':
                    pass


def hit(cur, delta, id):
    # if no row is updated, (id, delta) do not exists and insert
    cur.execute(
        "UPDATE limit_report SET hit=hit+1 WHERE id = %s and delta = %s" % (
            (config.format_str,)*2
        ),
        (id, delta)
    )
    if cur.rowcount <= 0:
        cur.execute(
            "INSERT INTO limit_report (id, delta, hit) VALUES (%s, %s, 1)" % (
                (config.format_str,)*2
            ),
            (id, delta)
        )


def write_pidfile():
    """write current pid file to ``config.pidfile``"""
    try:
        with open(config.pidfile, 'w') as f:
            f.write("%s" % os.getpid())
    except PermissionError as error:
        raise ValueError("Unable to write pid file, %s." % error)


def remove_pidfile():
    """Try to remove ``config.pidfile``"""
    try:
        os.remove(config.pidfile)
    except OSError:
        pass


def get_config(dotted_string):
    """
        Return the config parameter designated by ``dotted_string``.
        Dots are used as separator between dict and key.
    """
    params = dotted_string.split('.')
    obj = getattr(config, params[0])
    for param in params[1:]:
        obj = obj[param]
    return obj


config = LazyConfig()


policyd_rate_limit/policyd.py
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License version 3 for
# more details.
#
# You should have received a copy of the GNU General Public License version 3
# along with this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# (c) 2015-2016 Valentin Samir
import os
import sys
import socket
import time
import select

from policyd_rate_limit import utils
from policyd_rate_limit.utils import config


class Pass(Exception):
    pass


class Policyd(object):
    """The policy server class"""
    socket_data_read = {}
    socket_data_write = {}

    def socket(self):
        """initialize the socket from the config parameters"""
        # if socket is a string assume it is the path to an unix socket
        if isinstance(config.SOCKET, str):
            try:
                os.remove(config.SOCKET)
            except OSError:
                if os.path.exists(config.SOCKET):
                    raise
            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        # else asume its a tuple (bind_ip, bind_port)
        elif '.' in config.SOCKET[0]:  # assume ipv4 bind addresse
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        elif ':' in config.SOCKET[0]:  # assume ipv6 bind addresse
            sock = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
        else:
            raise ValueError("bad socket %s" % config.SOCKET)
        self.sock = sock

    def close_socket(self):
        """close the socket depending of the config parameters"""
        self.sock.close()
        # if socket was an unix socket, delete it after closing.
        if isinstance(config.SOCKET, str):
            try:
                os.remove(config.SOCKET)
            except OSError as error:
                sys.stderr.write("%s\n" % error)
                sys.stderr.flush()

    def close_connection(self, connection):
        """close a connection and clean read/write dict"""
        # Clean up the connection
        try:
            del self.socket_data_read[connection]
        except KeyError:
            pass
        try:
            del self.socket_data_write[connection]
        except KeyError:
            pass
        connection.close()

    def run(self):
        """The main server loop"""
        try:
            sock = self.sock
            sock.bind(config.SOCKET)
            if isinstance(config.SOCKET, str):
                os.chmod(config.SOCKET, config.socket_permission)
            sock.listen(1)
            self.socket_data_read[sock] = []
            if config.debug:
                sys.stderr.write('waiting for connections\n')
                sys.stderr.flush()
            while True:
                # wait for a socket to read to or to write to
                (rlist, wlist, _) = select.select(
                    self.socket_data_read.keys(), self.socket_data_write.keys(), []
                )
                for socket in rlist:
                    # if the socket is the main socket, there is a new connection to accept
                    if socket == sock:
                        connection, client_address = sock.accept()
                        if config.debug:
                            sys.stderr.write('connection from %s\n' % client_address)
                            sys.stderr.flush()
                        self.socket_data_read[connection] = []
                    # else there is data to read on a client socket
                    else:
                        self.read(socket)
                for socket in wlist:
                    try:
                        data = self.socket_data_write[socket]
                        sent = socket.send(data)
                        data_not_sent = data[sent:]
                        if data_not_sent:
                            self.socket_data_write[socket] = data_not_sent
                        else:
                            self.close_connection(socket)
                    # the socket has been closed during read
                    except KeyError:
                        pass
        except KeyboardInterrupt:
            for socket in self.socket_data_read.keys():
                if socket != self.sock:
                    self.close_connection(sock)
            raise

    def read(self, connection):
        """Called then a connection is ready for reads"""
        try:
            # get the current buffer of the connection
            buffer = self.socket_data_read[connection]
            # read data
            data = connection.recv(1024).decode('UTF-8')
            if not data:
                raise ValueError("connection closed")
            if config.debug:
                sys.stderr.write(data)
                sys.stderr.flush()
            # accumulate it in buffer
            buffer.append(data)
            # if data len too short to determine if we are on an empty line, we
            # concatene datas in buffer
            if len(data) < 2:
                data = u"".join(buffer)
                buffer = [data]
            # We reach on empty line so the client has finish to send and wait for a response
            if data[-2:] == "\n\n":
                data = u"".join(buffer)
                request = {}
                # read data are like one key=value per line
                for line in data.split("\n"):
                    line = line.strip()
                    try:
                        key, value = line.split(u"=", 1)
                        if value:
                            request[key] = value
                    # if value is empty, ignore it
                    except ValueError:
                        pass
                # process the collected data in the action method
                self.action(connection, request)
            else:
                self.socket_data_read[connection] = buffer
        except KeyboardInterrupt:
            self.close_connection(connection)
            raise
        except Exception as error:
            sys.stderr.write("%s\n" % error)
            sys.stderr.flush()
            self.close_connection(connection)

    def action(self, connection, request):
        """Called then the client has sent an empty line"""
        id = None
        # By default, we do not block emails
        action = config.success_action
        with utils.cursor() as cur:
            try:
                # if user is authenticated, we filter by sasl username
                if config.limit_by_sasl and u'sasl_username' in request:
                    id = request[u'sasl_username']
                # else, if activated, we filter by ip source addresse
                elif (
                    config.limit_by_ip and
                    u'client_address' in request and
                    utils.is_ip_limited(request[u'client_address'])
                ):
                    id = request[u'client_address']
                # if the client neither send us client ip adresse nor sasl username, jump
                # to the next section
                else:
                    raise Pass()
                # Here we are limiting agains sasl username or ip source addresses.
                # for each limit periods, we count the number of mails already send.
                # if the a limit is reach, we change action to fail (deny the mail).
                for mail_nb, delta in config.limits:
                    cur.execute(
                        (
                            "SELECT COUNT(*) FROM mail_count "
                            "WHERE id = %s AND date >= %s"
                        ) % ((config.format_str,)*2),
                        (id, int(time.time() - delta))
                    )
                    nb = cur.fetchone()[0]
                    if config.debug:
                        sys.stderr.write("%03d/%03d hit since %ss\n" % (nb, mail_nb, delta))
                        sys.stderr.flush()
                    if nb >= mail_nb:
                        action = config.fail_action
                        if config.report and delta in config.report_limits:
                            utils.hit(cur, delta, id)
                        raise Pass()
            except Pass:
                pass
            # If action is a success, record in the database that a new mail has just been sent
            if action == config.success_action and id is not None:
                if config.debug:
                    sys.stderr.write(u"insert id %s\n" % id)
                    sys.stderr.flush()
                cur.execute(
                    "INSERT INTO mail_count VALUES (%s, %s)" % ((config.format_str,)*2),
                    (id, int(time.time()))
                )
        data = u"action=%s\n\n" % action
        if config.debug:
            sys.stderr.write(data)
            sys.stderr.flush()
        # return the result to the client
        self.socket_data_write[connection] = data.encode('UTF-8')


policyd_rate_limit/config.py
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License version 3 for
# more details.
#
# You should have received a copy of the GNU General Public License version 3
# along with this program; if not, write to the Free Software Foundation, Inc., 51
# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# (c) 2015-2016 Valentin Samir
from policyd_rate_limit.const import SQLITE_DB, MYSQL_DB, PGSQL_DB  # noqa: F401

debug = True

user = "policyd-rate-limit"
group = "policyd-rate-limit"
pidfile = "/var/run/policyd-rate-limit/policyd-rate-limit.pid"

mysql_config = {
    "user": "username",
    "passwd": "*****",
    "db": "database",
    "host": "localhost",
    "charset": 'utf8',
}

sqlite_config = {
    "database": "/var/lib/policyd-rate-limit/db.sqlite3",
}

pgsql_config = {
    "database": "database",
    "user": "username",
    "password": "*****",
    "host": "localhost",
}

backend = SQLITE_DB

# SOCKET=("127.0.0.1", 8552)
SOCKET = "/var/spool/postfix/ratelimit/policy"
socket_permission = 0o666

# list of (number of mails, number of seconds)
limits = [
    (10, 60),  # limit to 10 mails by minutes
    (150, 86400),  # limits to 150 mails by days
]

limit_by_sasl = True
limit_by_ip = False

limited_netword = [
]

# actions return to postfix, see http://www.postfix.org/access.5.html for a list of actions.
success_action = "dunno"
fail_action = "defer_if_permit Rate limit reach, retry later"


# if True, send a report to report_email about users reaching limits each time --clean is called
report = False
# from who to send emails reports
report_from = None
# address to send emails reports to
report_to = None
# subject of the report email
report_subject = "policyd-rate-limit report"
# add number of seconds from the limits list for which you want to be reported
report_limits = [86400]
# only send a report if some users have reach a reported limit
report_only_if_needed = True

# The smtp server to use to send emails (host, port)
smtp_server = ("localhost", 25)
# Should we use starttls (you should set this to True if you use smtp_credentials)
smtp_starttls = False
# Should we use credentials to connect to smtp_server ? if yes set ("user", "password"), else None
smtp_credentials = None
